active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,821 @@
+persisted long term by stacks for use in future resource deletion.
+
+In the case where it's supplied by the user and without secret persistence, the only way the stacks service could
+possibly source this data again is to prompt the user for it. It is unclear how that experience would work. Does the
+user need to supply the entire old deployment's parameters back to the stack service? How do they know which version of
+inputs to supply? What if a subset of those parameters are no longer relevant? Would they only need to supply parameters
+used in configurations? How would we detect and differentiate that across an entire tree of deployments?
+
+In the case where it's supplied by a parent deployment, the problem becomes recursive. It's either provided by the user,
+is sourced from somewhere else in that parent deployment (such as another resource's outputs like a key vault or AKS
+cluster), or sourced from another parent deployment.
+
+To complicate this further, any pathway between the secret's source and consumption can have dynamic language
+expressions such as template function calls or branching. This would effectively require stacks to generate a separate
+deployment template at deletion time from the user's deployment template to process these language expressions.
+
+The stacks service cannot store user secrets, and it should not evaluate dynamic deployment template language
+expressions. The stacks service needs extension configurations to be defined in a format that is interpretable at
+deletion time that is transparent to the user and secure.
+
+### The new design from the Bicep perspective
+
+The following Bicep code illustrates the new design for extension configurations. The root deployment supplies the
+extension configuration details to the child deployment, allowing direct access to ARM resources in the root deployment
+that supply secrets.
+
+main.bicep - The root deployment
+
+```bicep
+param namespace string = 'default'
+
+// A sample key vault
+resource kv '...' = { } 
+
+// The ARM aksCluster resource can supply kubeConfigs for the AKS Kubernetes use case.
+resource aksCluster '...' = {
+  // ...
+}
+
+module extResources 'extResources.bicep' = {
+  name: 'extResourcesDeploymentName'
+  extensionConfigs: {
+    k8s: {
+      // a kubeConfig sourced from an ARM resource (AKS cluster):
+      kubeConfig: {
+        source: 'ResourceSecret'
+        resource: aksCluster
+        credentialType: 'Admin' // or user
+      }
+      // a kubeConfig sourced from a user expression (NOT allowed for stack deployments):
+      kubeConfig: { //
+        source: 'Custom'
+        value: aksCluster.listClusterAdminCredential().kubeconfigs[0].value 
+      } 
+      // a kubeConfig sourced from a Key Vault. Note: this will be compiled to the reference object format (source = 'KeyVault')
+      kubeConfig: kv.getSecret('myKubeConfig') // alternatively: { source: 'KeyVault' value: kv.getSecret(...) }
```

Comment:
From a user perspective, I like the `kv.getSecret('myKubeConfig')` syntax because it’s less verbose and aligns with how KV secrets are used in parameters. However, it introduces a few challenges:

- The user experience becomes inconsistent with the other two source types.
- Bicep doesn’t support union types where members can be either strings or objects with discriminators. For example, the following type definition results in an error:

![image](https://github.com/user-attachments/assets/1e8c0146-308d-41b6-9739-8d99022ac1f8)

The alternative approach `{ source: 'KeyVault', value: kv.getSecret('...') }` resolves the issues above, but it comes with added verbosity. Another drawback is that even with the `source: 'KeyVault'` discriminator, expression-level validation for value would still be necessary, as users could directly modify the generated ARM templates.

Given that expression validation cannot be avoided, a possible solution could involve leveraging credential source annotations to simplify expression analysis in both Bicep and the deployment engine.

## Source: Resource secret

Using the kubeConfig sourced from an AKS cluster as an example, instead of requiring users to explicitly specify an object discriminator, the extension provider could annotate the ResourceSecret source with metadata in the type manifest, such as:

- Allowed Resource Type: `'Microsoft.ContainerService/managedClusters'`
- Allowed List Operations: `listClusterAdminCredential` and `listClusterUserCredential` (but perhaps excluding `listClusterMonitoringUserCredential`?)
- 
This approach would allow Bicep to perform semantic validation based on the provided type metadata and enable users to write concise expressions for defining the kubeConfig value. E.g.,

```
{
  // Valid
  kubeConfig: aks.listClusterAdminCredential().kubeConfigs[0].value
  kubeConfig: aks.listClusterUserCredential().kubeConfigs[0].value

  // Invalid - Not an allowed list operation
  kubeConfig: aks.listMonitoringUserCredential().kubeConfigs[0].value

  // Invalid - Security risk sending a storage account key to the Kubernetes extension
  kubeConfig: storageAccount.listKeys().keys[0].value

  // Invalid - Any invalid expressions within the syntax tree are blocked
  kubeConfig: someCondition
    ? storageAccount.listKeys().keys[0].value
    : 'someValue'
}
```

The generated ARM template for the valid case would look like this:

```
{
  "kubeConfig": {
    "value": "[listClusterMonitoringUserCredential(resourceId('Microsoft.ContainerService/managedClusters', 'myAKSCluster'), '2024-08-01').kubeconfigs[0].value]"
  }
}
```

We would need to replicate the same semantic analysis performed on the Bicep client side within the deployment engine to block invalid list operation calls. This could be achieved using hard-coded extension facts.

The process should be straightforward, as we can walk through the expression syntax tree, scanning for list function calls to verify their validity. As long as we confirm that the value represents a valid AKS cluster list* property access expression, it can be safely stored for Stack.

## Source: Key Vault Secret

Specify a Key Vault reference would be very simple:

{
  kubeConfig: kv.getSecret('myKubeConfig')
}

The generated ARM template would contain:

```
{
  "kubeConfig": {
    "reference": {
        "keyVault": {
          "id": "/subscriptions/<subscription-id>/resourceGroups/<rg-name>/providers/Microsoft.KeyVault/vaults/<vault-name>"
        },
        "secretName": "myKubeConfig"
      }
  }
```

The handling of this in the deployment engine is trivial.

## Source: Custom

Any value that does not adhere to the format of a list operation or a KeyVault reference, or includes overly complex expressions, can be classified as a custom value.

- Bicep Behavior: Bicep can emit `information` (or `warning`) diagnostics to inform users that such values will not be supported for Stack deployments.
- Deployment Engine Behavior: During Stack deployments, the deployment engine can validate these values and throw a runtime error if they are encountered.

```
{
  // Explicit string value. Treated as a custom value.
  kubeConfig: 'someExplicitValue'

  // Reference to a parameter. Treated as a custom value.
  kubeConfig: someParameter

  // Reference to a variable. Treated as a custom value.
  kubeConfig: someVariable

  // Conditional expression with a valid list operation.
  // Supporting this would require extensive expression analysis, which is likely unnecessary.
  // Users should use conditional modules for such scenarios instead.
  kubeConfig: someVariable
    ? aks.listClusterAdminCredential().kubeConfig[0].value
    : 'someValueWhenFalse'

  // Conditional expression switching between two AKS resources.
  // Complex expression that is not recommended due to analysis overhead.
  kubeConfig: (someCondition ? aks1 : aks2).listClusterUserCredential().kubeConfig[0].value

  // Overly complex indexing logic with list operations.
  // Why would someone do this?
  kubeConfig: aks.listClusterUserCredential()[someParameter][someSecureParameter]
}
```

The key here is to streamline maintenance by minimizing expression analysis. This approach avoids partial expression reduction and sidesteps obscure edge cases. Here's a straightforward processing flow for the deployment engine:

1. Security Check: Determine if the extension configuration item is secure using extension facts.

  - If not secure, it can be safely saved for Stack.

2. KeyVault Reference Validation: If the item is secure, check whether it is a KeyVault reference.

  - If it is a KeyVault reference, save it for Stack.

3. List Operation Validation: If not a KeyVault reference, scan for invalid list operations based on extension facts.

  - If it matches a valid list* property access expression and it does not contain any weird parameter / variable reference or condition expressions, save it for Stack.

4. Custom Value Handling: For any other cases that do not fit the above criteria:

  - Treat the value as a custom configuration.
  - Return an error for Stack deployments.

I believe the above solution offers several key benefits. It provides a straightforward and consistent approach for users to specify configuration items, minimizing confusion. Furthermore, it simplifies the complexity of extension facts - rather than storing detailed instructions for interpreting configuration items to retrieve credentials, it only requires tracking the resource type and a list of valid list operations. This approach is also flexible enough to be generalized for other extensions in the future.

Reply:
About `kv.getSecret()`: this currently does compile to a key vault reference in the ARM template because it's only usable for secure nested deployment parameters so we wouldn't be unioning a string here. I was wondering if we should try to make it consistent to the nested deployment parameter style but I do agree it produces an inconsistency with the other sources. There's a couple choices on the Bicep side here:
1. Go full object literal for references, user must write out the object by hand.
2. Provide Bicep-only functions that compile to the reference format. This allow "Custom" expressions without the object wrapper and would require a special function for the "Resource" source (ex `resourceSecret(aks, { credentialType: '' }`) although this would more or equally be difficult to implement compared to the object literal due to the introduction of properties determined by the resource passed in). Maybe this can be worked around with your "resource secret" suggestion.

# Source: Resource secret
I'm wondering what the user experience will be with this. 

How will we guide the user to write an expression based on dynamic resource type(s) and with it, dynamic list* calls? I suppose there can be a couple diagnostics here, 1 for the expectation of a resource list* call of a set of resource types, and then a diagnostic for the list* part of the expression that lists out the acceptable ones. That paired with some examples in the documentation is probably enough.

We lose the extension guided value resolution with the object literal format and it off loads that logic back to the user. For example, take the AKS Admin vs User credentials case. With the literal format, the user could declare a strategy instead of a credentialType where the extension decides the right credential to use depending on information supplied to it, either from the template, deployment engine (submitting user for instance), or the configuration of the resource (the resource inputs, and any alteration some deployment script may make to it). I'm not sure if this specific example is an actual use case, but I think having the layer of `(deployment, template, resource) => credentialInstructions` in the extension would be beneficial to have. Do we want to give this up?

# Source: Key Vault Secret
I think we should go with this because this is the same format for deployment parameters.

# Source: Custom
Seems viable. We can check the inner most function. If it's a list* and there's no 'if's along the way we assume the resource source, otherwise custom.

Reply:
> About kv.getSecret(): this currently does compile to a key vault reference in the ARM template because it's only usable for secure nested deployment parameters so we wouldn't be unioning a string here

It compiles into an object in the ARM template; however, the return type of kv.getSecret() in Bicep is string. When specifying kubeConfig in Bicep, we must choose either object format or string expressions, but mixing both will not be supported.

Reply:
> I suppose there can be a couple diagnostics here, 1 for the expectation of a resource list* call of a set of resource types, and then a diagnostic for the list* part of the expression that lists out the acceptable ones. That paired with some examples in the documentation is probably enough.

That’s along the lines of what I was thinking. We could also enhance the experience by incorporating type-guided completions.

> We lose the extension guided value resolution with the object literal format and it off loads that logic back to the user.

Agreed. It does compromise flexibility, which is something I’d prefer not to lose. Providing users with a literal `Admin | User` credential type also feels like a much better experience overall.

I wonder if there’s a solution that could combine the strengths of both approaches, offering a concise and user-friendly experience while preserving flexibility...

Reply:
I have another idea! It still uses an object discriminator but avoids the caveats of union types:

```
{
  k8s: {
    // Descriminator for clusterType rather than kubeConfig
    clusterType: 'Managed | Custom'

    // clusterType == 'Managed'
    // Extension-guided resolution is applied.
    managedCluster: {
      id: aks.Id
      credentialType: 'Admin' | 'User'.
    }
    
    // clusterType == 'Custom'
    // KV reference - supported by Stacks.
    kubeConfig: kv.getSecret('kubeConfig')

    // Non-KV reference - not supported by Stacks.
    kubeConfig: someParam // Example of unsupported value
    kubeConfig: whatever  // Example of unsupported value
  }
}
```

What do you think?

Reply:
> I have another idea! It still uses an object discriminator but avoids the caveats of union types:
> 
> ```bicep
> {
>   k8s: {
>     // Descriminator for clusterType rather than kubeConfig
>     clusterType: 'Managed | Custom'
> 
>     // clusterType == 'Managed'
>     // Extension-guided resolution is applied.
>     managedClusterId: aks.id
>     credentialType: 'User' | 'Admin' // Optional. Default is 'User'.
>     
>     // clusterType == 'Custom'
>     // KV reference - supported by Stacks.
>     kubeConfig: kv.getSecret('kubeConfig')
> 
>     // Non-KV reference - not supported by Stacks.
>     kubeConfig: someParam
>     kubeConfig: whatever
>   }
> }
> ```
> 
> What do you think?
This seems like a good alternative. I'll work through the details of this see if there's any issues.